aws/src/test/java/org/apache/iceberg/aws/s3/TestFuzzyS3InputStream.java

```diff
@@ -0,0 +1,224 @@
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
+
+public class TestFuzzyS3InputStream extends TestS3InputStream {
+
+  private static final int DATA_SIZE = 100;
+  private static final int SEEK_SIZE = 4;
+  private static final int SEEK_NEW_POSITION = 25;
+
+  @ParameterizedTest
+  @MethodSource("retryableExceptions")
+  public void testReadWithFuzzyStreamRetrySucceed(IOException exception) throws Exception {
+    testRead(fuzzyStreamClient(new AtomicInteger(3), exception), DATA_SIZE);
+  }
+
+  @ParameterizedTest
+  @MethodSource("retryableExceptions")
+  public void testReadWithFuzzyStreamExhaustedRetries(IOException exception) {
+    assertThatThrownBy(
+            () -> testRead(fuzzyStreamClient(new AtomicInteger(5), exception), DATA_SIZE))
+        .isInstanceOf(exception.getClass())
+        .hasMessage(exception.getMessage());
+  }
+
+  @ParameterizedTest
+  @MethodSource("nonRetryableExceptions")
+  public void testReadWithFuzzyStreamNonRetryableException(IOException exception) {
+    assertThatThrownBy(
+            () -> testRead(fuzzyStreamClient(new AtomicInteger(3), exception), DATA_SIZE))
+        .isInstanceOf(exception.getClass())
+        .hasMessage(exception.getMessage());
+  }
+
+  @Override
+  protected void testRead(S3Client s3, int dataSize) throws Exception {
+    testRead(s3, DATA_SIZE, 4, SEEK_SIZE, SEEK_NEW_POSITION);
+  }
+
+  private static Stream<Arguments> retryableExceptions() {
+    return Stream.of(
+        Arguments.of(
+            new SocketTimeoutException("socket timeout exception"),
+            new SSLException("some ssl exception")));
+  }
+
+  private static Stream<Arguments> nonRetryableExceptions() {
+    return Stream.of(Arguments.of(new IOException("some generic non-retryable IO exception")));
+  }
+
+  private S3ClientWrapper fuzzyStreamClient(AtomicInteger counter, IOException failure) {
+    S3ClientWrapper fuzzyClient = spy(new S3ClientWrapper(s3Client()));
+    doAnswer(
+            invocation ->
+                new FuzzyResponseInputStream(invocation.callRealMethod(), counter, failure))
+        .when(fuzzyClient)
+        .getObject(any(GetObjectRequest.class), any(ResponseTransformer.class));
+    return fuzzyClient;
+  }
+
+  /** Wrapper for S3 client, used to mock the final class DefaultS3Client */
+  public static class S3ClientWrapper implements S3Client {
+
+    private final S3Client delegate;
+
+    public S3ClientWrapper(S3Client delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public String serviceName() {
+      return delegate.serviceName();
+    }
+
+    @Override
+    public void close() {
+      delegate.close();
+    }
+
+    @Override
+    public <ReturnT> ReturnT getObject(
+        GetObjectRequest getObjectRequest,
+        ResponseTransformer<GetObjectResponse, ReturnT> responseTransformer)
+        throws AwsServiceException, SdkClientException {
+      return delegate.getObject(getObjectRequest, responseTransformer);
```

Comment:
I was going to track the start ranges that were requested on each getObject and assert that the positions being retried are as expected but this ends up really complicating the tests for what I think is not too much additional value because actually the `testRead` we are inheriting verifies the contents actually being read in each `readAndCheck`. If there's a retry that's being performed to an incorrect position, this would mean that the contents would be incorrect and tests would fail and we know that the retry path is being exercised given the presence of the fuzzy stream client.